src/Inciter/AMR/marked_refinements_store.h

```diff
@@ -0,0 +1,113 @@
+#ifndef AMR_marked_refinements_store_h
+#define AMR_marked_refinements_store_h
+
+#include <map>
+#include "Refinement_State.h"
+
+namespace AMR {
+
+    class marked_refinements_store_t {
+        private:
+            std::map<size_t, Refinement_Case> marked_refinements;
+
+            // TODO: This probably isn't the right place for this
+            // We will use this variable to check if anything has changed
+            // during a round of refinement marking
+            bool refinement_state_changed = false;
+
+        public:
+            /**
+             * @brief function to see if a given id has already had a
+             * refinement decision made
+             *
+             * @param id The tet_id to check
+             *
+             * @return A bool stating if a refinement decision/marking exists
+             */
+            bool exists(size_t id)
+            {
+                auto f = marked_refinements.find(id);
+                if (f != marked_refinements.end())
+                {
+                    trace_out << "Marked element " << id << " has value " <<
+                        f->second << std::endl;
+
+                    return true;
+                }
+                return false;
+            }
+
+            /**
+             * @brief Accessor function to get a marked_refinement from the store
+             *
+             * @param id The id of the tet to fetch the marked_refinement for
+             *
+             * @return The marked refinement case for the given tet
+             */
+            Refinement_Case& get(size_t id)
+            {
+                return marked_refinements.at(id);
+            }
+
+            void erase(size_t id)
+            {
+                marked_refinements[id] = Refinement_Case::none;
+            }
+
+            /**
+             * @brief Function to add a marked refinement to the store, which
+             * is context aware based on existing values
+             *
+             * @param id The id of the tet to mark
+             * @param r The refinement decision for the tet
+             */
+            void add(size_t id, Refinement_Case r)
+            {
+                // Check if that active element already exists
+                if (exists(id))
+                {
+                    if (marked_refinements[id] != r)
+                    {
```

Comment:
Nice idea, but to me it very much feels like a premature "optimization". I think we would do better to just have stronger semantics around what is stored. 

At a minimum, we could just probably remove the `exists` call, or replace it with a `find` call.  

I think we can punt on this, and then if we decide to address the global nature of `refinement_state_changed` (imo a bigger problem), this will be natural to address then